pkg/kube_events_manager/monitor_test.go

```diff
@@ -0,0 +1,143 @@
+package kube_events_manager
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/flant/kube-client/fake"
+	"github.com/flant/kube-client/manifest"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	. "github.com/flant/shell-operator/pkg/kube_events_manager/types"
+)
+
+func Test_Monitor_should_handle_dynamic_ns_events(t *testing.T) {
+	g := NewWithT(t)
+	fc := fake.NewFakeCluster(fake.ClusterVersionV121)
+
+	// Initial namespace with ConfigMap.
+	createNsWithLabels(fc, "default", map[string]string{"test-label": ""})
+	createCM(fc, "default", testCM("default-cm-1"))
+
+	monitorCfg := &MonitorConfig{
+		ApiVersion: "v1",
+		Kind:       "ConfigMap",
+		EventTypes: []WatchEventType{WatchEventAdded, WatchEventModified, WatchEventDeleted},
+		NamespaceSelector: &NamespaceSelector{
+			LabelSelector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"test-label": "",
+				},
+			},
+		},
+	}
+
+	mon := NewMonitor()
+	mon.WithContext(context.TODO())
+	mon.WithKubeClient(fc.Client)
+	mon.WithConfig(monitorCfg)
+
+	// Catch KubeEvents from monitor.
+	type kubeEvent struct {
+		Type  KubeEventType
+		Count int
+	}
+	events := make([]kubeEvent, 0)
+	mon.WithKubeEventCb(func(ev KubeEvent) {
+		events = append(events, kubeEvent{
+			Type:  ev.Type,
+			Count: len(ev.Objects),
+		})
+	})
+
+	// Start monitor.
+	err := mon.CreateInformers()
+	g.Expect(err).ShouldNot(HaveOccurred())
+	mon.Start(context.TODO())
+
+	// Get initial snapshot.
+	snap := mon.Snapshot()
+	g.Expect(snap).Should(HaveLen(1), "Should have only one ConfigMap on start")
+	g.Expect(events).Should(HaveLen(0), "Should not call KubeEventCb until EnableKubeEventsCb")
+
+	// Simulate object creation during Synchronization phase:
+	// create new ns with matching labels and then create new ConfigMap.
+	createNsWithLabels(fc, "test-ns-1", map[string]string{"test-label": ""})
+
+	// Wait until informer appears.
+	g.Eventually(func() bool {
+		monImpl := mon.(*monitor)
+		return len(monImpl.VaryingInformers) == 2
```

Comment:
Counting seems good enough since this behavior looks deterministic. But I suggest to be more strict and test that `VaryingInformers` actually contains expected keys